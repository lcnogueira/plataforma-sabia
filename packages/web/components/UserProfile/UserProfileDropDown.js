import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import LogoutButton from './LogoutButton';
import PageLink from './PageLink';
import sections from './sections';

const UserDropDown = ({ visible }) => {
	const { t } = useTranslation(['profile']);

	return (
		visible && (
			<DropDownContainer>
				<DropDownMenu>
					{sections.map(({ pages }) =>
						pages.map((page) => (
							<PageLink key={page.slug} href={page.href}>
								<page.icon />
								{t(page.slug)}
							</PageLink>
						)),
					)}
					<div>
						<LogoutButton />
					</div>
				</DropDownMenu>
			</DropDownContainer>
		)
	);
};

UserDropDown.propTypes = {
	visible: PropTypes.bool,
};

UserDropDown.defaultProps = {
	visible: false,
};

const DropDownContainer = styled.div`
	position: relative;
`;

const DropDownMenu = styled.div`
	${({ theme: { colors } }) => css`
		position: absolute;
		width: 24rem;
		left: calc(50% - 13rem);
		top: calc(100% + 3rem);
		background: ${colors.white};
		border-radius: 0.4rem;
		padding: 2rem;
		box-shadow: 0 2px 4px 0 ${colors.secondary};
		transition: 0.3s;

		:hover {
			box-shadow: 0 4px 8px 0 ${colors.secondary};
		}

		> div {
			padding-top: 1rem;
			border-top: 0.1rem solid ${colors.border};
		}

		&::before {
			content: '';
			position: absolute;
			left: calc(50% - 2rem);
			top: -2rem;
			width: 0;
			height: 0;
			border-left: 2rem solid transparent;
			border-right: 2rem solid transparent;
			border-bottom: 2rem solid ${colors.white};
		}
	`}
`;

export default UserDropDown;
