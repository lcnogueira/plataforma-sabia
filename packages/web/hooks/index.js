// eslint-disable-next-line import/prefer-default-export
export { default as useModal } from './useModal';
