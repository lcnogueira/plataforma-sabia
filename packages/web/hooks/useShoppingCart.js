import { useContext } from 'react';
import ShoppingCartContext from '../components/ShoppingCart/ShoppingCartContext';

function useShoppingCart() {
	const context = useContext(ShoppingCartContext);
	if (context === undefined) {
		throw new Error('useShoppingCart must be used within a ShoppingCartProvider');
	}
	return context;
}

export default useShoppingCart;
