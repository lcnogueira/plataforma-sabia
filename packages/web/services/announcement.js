/* eslint-disable import/prefer-default-export */
import { apiPost } from './api';

/**
 * Creates an annonucement.
 *
 * @param {object} announcementData The announcement data
 * @returns {object} The newly created announcement
 */
export const createAnnouncement = async (announcementData) => {
	const { data, status } = await apiPost(`announcements`, { ...announcementData });

	return status !== 200 ? { data: data?.error, success: false } : { data, success: true };
};
