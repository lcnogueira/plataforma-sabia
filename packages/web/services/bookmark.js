/* eslint-disable import/prefer-default-export */
import { apiPost, apiDelete } from './api';

/**
 * Normalize handleBookmark request
 *
 * @typedef {object} HandleBookmarkRequest
 * @property {boolean} [params.active=true] Current bookmark state
 * @property {number} params.technologyId The technology id
 * @property {string} params.userToken The JWT token
 * @property {number} params.userId The user id
 *
 * @param {HandleBookmarkRequest} params Bookmark params
 * @returns {object} The newly bookmark.
 */
const parseHandleBookmarkRequest = ({ active = true, technologyId, userToken, userId }) => {
	let method;
	let endpoint;

	if (active) {
		method = apiDelete;
		endpoint = `user/${userId}/bookmarks`;
	} else {
		method = apiPost;
		endpoint = 'bookmarks';
	}

	return {
		method,
		endpoint,
		technologyIds: [technologyId],
		userToken,
	};
};

/**
 * Create or delete an user bookmark.
 *
 * @param {HandleBookmarkRequest} params Bookmark params
 * @returns {object} The newly bookmark
 */
export const handleBookmark = async (params) => {
	const { method, endpoint, technologyIds, userToken } = parseHandleBookmarkRequest(params);

	const response = await method(endpoint, {
		userToken,
		technologyIds,
	});

	return response.status === 200;
};
