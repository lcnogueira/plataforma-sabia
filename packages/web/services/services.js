import { apiGet, apiPost } from './api';

/**
 * GETs all services
 *
 * @param {object} options Optional params sent in the request
 * @returns {Array} The services ata
 */
export const getServices = async (options = {}) => {
	const response = await apiGet('services', { embed: true, ...options });

	if (response.status !== 200) {
		return [];
	}

	return response.data;
};

/**
 * Creates a service order
 *
 * @param {Array} servicesToCreate Array of services to create order
 * @returns {object} The services that was bought
 */
export const createServiceOrder = async (servicesToCreate) => {
	const response = await apiPost('services/orders', { services: servicesToCreate });

	if (response.status !== 200) {
		return false;
	}

	return response.data;
};
