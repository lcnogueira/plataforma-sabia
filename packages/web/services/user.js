/* eslint-disable import/prefer-default-export */
import { apiGet, apiPut, apiPost } from './api';

/**
 * Updates the password of a logged in user.
 *
 * @param {object} data User data
 * @param {object} data.currentPassword The current password
 * @param {object} data.newPassword The new password
 *
 * @returns {object} The success message or an error object.
 */
export const updateUserPassword = async ({ currentPassword, newPassword }) => {
	const response = await apiPut(`user/change-password`, { currentPassword, newPassword });

	return response.data;
};

/**
 * Updates the email of a logged in user.
 *
 * @param {string} newEmail The new email
 * @returns {object} The success message or an error object.
 */
export const requestEmailChange = async (newEmail) => {
	const response = await apiPost(`user/change-email`, { email: newEmail, scope: 'web' });

	return response.data;
};

/**
 * Updates an existing user.
 *
 * @param {number} id The id of the user to update
 * @param {object} data The user data.
 *
 * @returns {object} The updated user.
 */
export const updateUser = async (id, data) => {
	if (!id) {
		return false;
	}

	const response = await apiPut(`users/${id}`, data);

	return response.data;
};

/**
 * Sends a request to ask for reviewer permission.
 *
 * @param {number} id The id of the user
 * @param {object} categories The categories to be a reviewer
 *
 * @returns {object} The updated user.
 */
export const requestToBeReviewer = async (id, { categories } = {}) => {
	if (!id || !categories) {
		return false;
	}

	const response = await apiPost('reviewers', {
		user_id: id,
		categories,
		disclaimers: Array.from(Array(15).keys()),
	});

	if (response.status !== 200) return false;

	return response.data;
};

/**
 * Fetches technologies of a given user.
 *
 * @param {number} userId The user id.
 * @param {object} options Optional params.
 */
export const getUserTechnologies = async (userId, options = { embed: true }) => {
	const response = await apiGet(`users/${userId}`, { embed: options.embed });

	if (response.status !== 200) {
		return false;
	}

	const { data } = response;

	return data.technologies;
};

/**
 * Fetches favorite technologies of a given user.
 *
 * @param {number} userId The user id.
 * @param {object} options Optional params
 */
export const getUserBookmarks = async (userId, options = { embed: true }) => {
	const response = await apiGet(`user/${userId}/bookmarks`, { ...options, embed: true });

	if (response.status !== 200) {
		return false;
	}

	const { data, headers } = response;

	const totalPages = headers['X-Sabia-Total-Pages'];
	const totalItems = headers['X-Sabia-Total'];

	return { data, totalPages, totalItems };
};

/**
 * Fetches favorite technologies of a given user.
 *
 * @param {object} options Optional params
 */
export const getReviewerUser = async (options = { embed: false }) => {
	const response = await apiGet(`reviewer`, { ...options });

	if (response.status !== 200) {
		return false;
	}

	const { data } = response;

	return { data };
};

/**
 * Fetches questions of a given user.
 *
 * @param {object} options Optional params
 */
export const getUserQuestions = async (options = { embed: true }) => {
	const response = await apiGet(`questions`, { ...options, embed: true });

	if (response.status !== 200) {
		return false;
	}

	const { data, headers } = response;

	const totalPages = headers['X-Sabia-Total-Pages'];
	const totalItems = headers['X-Sabia-Total'];

	return { data, totalPages, totalItems };
};

/**
 * Fetches the count of unanswered questions of a given user.
 */
export const getUserUnansweredQuestions = async () => {
	const response = await apiGet(`questions/unanswered`);

	if (response.status !== 200) {
		return false;
	}

	const { data } = response;

	return { data };
};
