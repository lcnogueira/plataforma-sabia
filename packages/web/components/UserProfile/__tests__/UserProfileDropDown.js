import React from 'react';
import { render, screen, act } from 'test-utils';
import { cache } from 'swr';
import UserProfileDropDown from '../UserProfileDropDown';
import * as useAuth from '../../../hooks/useAuth';
import * as userServices from '../../../services/user';

describe('UserProfileDropDown', () => {
	afterEach(() => {
		cache.clear();
	});

	test('it renders as expected if visible is true', () => {
		const { container } = render(<UserProfileDropDown visible toggleVisible={() => {}} />);

		expect(screen.queryAllByRole('list').length).toBeGreaterThan(0);
		expect(screen.queryByRole('button', { name: /sair|logout/i })).toBeTruthy();
		expect(container).toMatchSnapshot();
	});

	test('it does not render if receives visible as false', () => {
		const { container } = render(
			<UserProfileDropDown visible={false} toggleVisible={() => {}} />,
		);

		expect(screen.queryAllByRole('list')).toHaveLength(0);
		expect(screen.queryByRole('button', { name: /sair|logout/i })).toBeNull();
		expect(container).toMatchSnapshot();
	});

	it('should not request notifications if user is unauthenticated', () => {
		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {},
		});
		const mockedQuestions = jest.spyOn(userServices, 'getUserUnansweredQuestions');
		const mockedMessages = jest.spyOn(userServices, 'getUserNewMessages');

		render(<UserProfileDropDown visible={false} toggleVisible={() => {}} />);

		expect(mockedQuestions).toHaveBeenCalledTimes(0);
		expect(mockedMessages).toHaveBeenCalledTimes(0);
	});

	it('should request notifications if user is authenticated', async () => {
		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {
				email: 'test@email.com',
			},
			isAuthenticated: true,
		});
		const mockedQuestions = jest.spyOn(userServices, 'getUserUnansweredQuestions');
		const mockedMessages = jest.spyOn(userServices, 'getUserNewMessages');

		act(() => {
			render(<UserProfileDropDown visible toggleVisible={() => {}} />);
		});

		// This expect is only for waiting the new render of SWR so we avoid ACT warnings
		expect(await screen.findByRole('button')).toBeInTheDocument();
		expect(mockedQuestions).toHaveBeenCalledTimes(1);
		expect(mockedMessages).toHaveBeenCalledTimes(1);
	});
});
