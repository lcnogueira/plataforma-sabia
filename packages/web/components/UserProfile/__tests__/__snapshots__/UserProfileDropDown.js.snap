// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileDropDown it does not render if receives visible as false 1`] = `<div />`;

exports[`UserProfileDropDown it renders as expected if visible is true 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: 1.6rem;
  background: none;
  border: none;
  color: hsla(0,79%,63%);
  padding-left: 2rem;
}

.c3:hover {
  color: hsla(0,58%,57%);
}

.c3 svg {
  fill: hsla(0,79%,63%);
  stroke: hsla(0,79%,63%);
  width: 2rem;
  height: 2rem;
  margin-right: 1rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.4rem;
  font-weight: 500;
  margin-bottom: 2rem;
  color: hsla(0,0%,52%);
  padding-left: 2rem;
}

.c2 svg {
  margin-right: 1rem;
}

.c2:hover {
  color: hsla(146,64%,36%);
}

.c0 {
  position: relative;
}

.c1 {
  position: absolute;
  min-width: 24rem;
  left: calc(50% - 11rem);
  top: calc(100% + 2.1rem);
  background: hsla(0,0%,100%);
  border-radius: 0.5rem;
  padding: 2rem 1.6rem;
  box-shadow: 0 0 2rem -1.5rem hsl(166,100%,34%);
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c1 li a,
.c1 li button {
  padding: 0rem;
  font-size: 1.6rem;
  font-weight: 400;
}

.c1:hover {
  box-shadow: 0 0 2.2rem -1.5rem hsl(166,100%,34%);
}

.c1::before {
  content: '';
  position: absolute;
  left: calc(50% - 2rem);
  top: -2rem;
  width: 0;
  height: 0;
  border-left: 2rem solid transparent;
  border-right: 2rem solid transparent;
  border-bottom: 2rem solid hsla(0,0%,100%);
}

.c1 li:last-child {
  padding-top: 1rem;
  border-top: 0.1rem solid hsla(0,0%,93%);
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 21v-2a4 4 0 00-4-4H8a4 4 0 00-4 4v2M12 11a4 4 0 100-8 4 4 0 000 8z"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:myProfile
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M22 6l-10 7L2 6"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:messages
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="20"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.84 2.61a5.5 5.5 0 00-7.78 0L12 3.67l-1.06-1.06a5.501 5.501 0 00-7.78 7.78l1.06 1.06L12 19.23l7.78-7.78 1.06-1.06a5.5 5.5 0 000-7.78v0z"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:favoriteTechnologies
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 2L3 6v14a2 2 0 002 2h14a2 2 0 002-2V6l-3-4H6zM3 6h18M16 10a4 4 0 11-8 0"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:userOrders
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 12h-6l-2 3h-4l-2-3H2"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M5.45 5.11L2 12v6a2 2 0 002 2h16a2 2 0 002-2v-6l-3.45-6.89A2 2 0 0016.76 4H7.24a2 2 0 00-1.79 1.11v0z"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:myTechnologies
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6H21"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M8 12H21"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M8 18H21"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M3 6H3.01"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M3 12H3.01"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M3 18H3.01"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:myServices
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 4h2a2 2 0 012 2v14a2 2 0 01-2 2H6a2 2 0 01-2-2V6a2 2 0 012-2h2"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M15 2H9a1 1 0 00-1 1v2a1 1 0 001 1h6a1 1 0 001-1V3a1 1 0 00-1-1z"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:orders
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          <svg
            fill="none"
            height="24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10z"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
            <path
              d="M9.09 9a3 3 0 015.83 1c0 2-3 3-3 3M12 17h.01"
              stroke="hsl(166, 100%, 34%)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          profile:questions
        </a>
      </li>
      <li>
        <button
          class="c3"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M868 732h-70.3c-4.8 0-9.3 2.1-12.3 5.8-7 8.5-14.5 16.7-22.4 24.5a353.84 353.84 0 0 1-112.7 75.9A352.8 352.8 0 0 1 512.4 866c-47.9 0-94.3-9.4-137.9-27.8a353.84 353.84 0 0 1-112.7-75.9 353.28 353.28 0 0 1-76-112.5C167.3 606.2 158 559.9 158 512s9.4-94.2 27.8-137.8c17.8-42.1 43.4-80 76-112.5s70.5-58.1 112.7-75.9c43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 7.9 7.9 15.3 16.1 22.4 24.5 3 3.7 7.6 5.8 12.3 5.8H868c6.3 0 10.2-7 6.7-12.3C798 160.5 663.8 81.6 511.3 82 271.7 82.6 79.6 277.1 82 516.4 84.4 751.9 276.2 942 512.4 942c152.1 0 285.7-78.8 362.3-197.7 3.4-5.3-.4-12.3-6.7-12.3zm88.9-226.3L815 393.7c-5.3-4.2-13-.4-13 6.3v76H488c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h314v76c0 6.7 7.8 10.5 13 6.3l141.9-112a8 8 0 0 0 0-12.6z"
            />
          </svg>
          profile:logout
        </button>
      </li>
    </ul>
  </div>
</div>
`;
